src/docs/eos-autoupdater.8
.\" Manpage for eos-autoupdater.
.\" Documentation is under the same licence as the eos-updater package.
.TH man 8 "28 Feb 2017" "1.0" "eos\-autoupdater man page"
.\"
.SH NAME
.IX Header "NAME"
eos\-autoupdater — Endless OS Automatic Updater
.\"
.SH SYNOPSIS
.IX Header "SYNOPSIS"
.\"
\fBeos\-autoupdater [\-\-force\-update] [\-\-from\-volume \fPpath\fB]\fP
.\"
.SH DESCRIPTION
.IX Header "DESCRIPTION"
.\"
\fBeos\-autoupdater\fP is a service which couples with \fBeos\-updater\fP(8) to
automate downloading and applying updates. While \fBeos\-updater\fP(8) provides
the implementation of polling for, fetching, and applying updates;
\fBeos\-autoupdater\fP provides the policy of when to poll, and how many steps
of the process should be done automatically. It controls the D\-Bus API exposed
by \fBeos\-updater\fP(8).
.PP
\fBeos\-autoupdater\fP is designed to be run by \fBsystemd\fP(1) on a timer
(the \fBeos\-autoupdater.timer\fP unit) which is set to poll for updates around
once an hour. \fBeos\-autoupdater\fP should not typically be run manually, but
can be triggered by starting the \fBeos\-autoupdater.service\fP unit. This will
trigger a poll for updates (and it potentially fetches and applies any which
are found). See \fBsystemctl\fP(1).
.PP
To interact with \fBeos\-updater\fP more directly, use the
\fBeos\-updater\-ctl\fP(8) tool.
.PP
Updates are retrieved in several stages; see \fBeos\-updater\fP(8). The
autoupdater controls the transitions between these stages, automatically
advancing as far as it can up to (and including) the stage specified by
\fILastAutomaticStep=\fP in \fBeos\-autoupdater.conf\fP(5), but no further.
The user must manually trigger further stages.
.PP
The \fBeos\-autoupdater.conf\fP(5) configuration file allows several policies
to be specified to control when updates are polled for. Use the
\fB\-\-force\-update\fP option to ignore these policies and unconditionally
poll for an update.
.PP
To apply Endless OS updates from a USB drive which is not configured in
\fBeos\-updater.conf\fP(5), use the \fB\-\-from\-volume\fP option to provide
the USB drive mount path. By default, updates on USB drives will be ignored.
Use \fBeos\-updater\-prepare\-volume\fP(8) to put updates onto a USB drive.
.PP
If the computer has been converted to not use OSTree, automatic updates are
permanently disabled.
.\"
.SH OPTIONS
.IX Header "OPTIONS"
.\"
.IP "\fB\-\-force\-update\fP"
Force the updater to poll for an update (and potentially fetch and apply it),
regardless of the policies configured in \fBeos\-autoupdater.conf\fP(5).
.\"
.IP "\fB\-\-from\-volume=\fP"
Path to a mounted USB drive which contains Endless OS updates prepared using
\fBeos\-updater\-prepare\-volume\fP(8). Force the updater to poll for updates
from the USB drive, and fetch and apply them locally if they are more recent
than the booted operating system.
.\"
.SH "ENVIRONMENT"
.IX Header "ENVIRONMENT"
.\"
\fPeos\-autoupdater\fP supports the standard GLib environment variables for
debugging. These variables are \fBnot\fP intended to be used in production:
.\"
.IP \fI$G_MESSAGES_DEBUG\fP 4
.IX Item "$G_MESSAGES_DEBUG"
This variable can contain one or more debug domain names to display debug output
for. The value \fIall\fP will enable all debug output. The default is for no
debug output to be enabled.
.\"
.SH "EXIT STATUS"
.IX Header "EXIT STATUS"
.\"
\fBeos\-autoupdater\fP may return one of several error codes if it encounters
problems.
.\"
.IP "0" 4
.IX Item "0"
No problems occurred. The updater ran successfully, and either polled for
updates, or determined not to based on policies configured in
\fBeos\-autoupdater.conf\fP(5).
.\"
.IP "1" 4
.IX Item "1"
There was a D\-Bus error while running the updater, or the updater entered an
error state after a poll, fetch or apply operation.
.\"
.IP "2" 4
.IX Item "2"
An invalid option was passed to \fBeos\-autoupdater\fP on startup.
.\"
.IP "3" 4
.IX Item "3"
The configuration file for \fBeos\-autoupdater\fP could not be loaded or
parsed.
.\"
.SH "FILES"
.IX Header "FILES"
.\"
.IP \fI/etc/eos\-updater/eos\-autoupdater.conf\fP 4
.IX Item "/etc/eos\-updater/eos\-autoupdater.conf"
.IX Item "/usr/local/share/eos\-updater/eos\-autoupdater.conf"
.IX Item "/usr/share/eos\-updater/eos\-autoupdater.conf"
Configuration file for \fBeos\-autoupdater\fP. Defaults are loaded from
\fI/usr/local/share/eos\-updater/eos\-autoupdater.conf\fP (if present) or
\fI/usr/share/eos\-updater/eos\-autoupdater.conf\fP. See
\fBeos\-autoupdater.conf\fP(5).
.\"
.IP \fI/lib/systemd/system/eos\-autoupdater.timer\fP 4
.IX Item "/lib/systemd/system/eos\-autoupdater.timer"
\fBsystemd\fP(1) timer file which triggers \fBeos\-autoupdater\fP to run
at regular intervals. See \fBsystemd.timer\fP(5).
.\"
.IP \fI/lib/systemd/system/eos\-autoupdater.service\fP 4
.IX Item "/lib/systemd/system/eos\-autoupdater.service"
\fBsystemd\fP(1) service file which specifies the runtime environment for
\fBeos\-autoupdater\fP. See \fBsystemd.service\fP(5).
.\"
.SH "SEE ALSO"
.IX Header "SEE ALSO"
.\"
\fBeos\-autoupdater.conf\fP(5),
\fBeos\-updater\fP(8),
\fBeos\-updater\-ctl\fP(8),
\fBeos\-updater\-prepare\-volume\fP(8)
.\"
.SH BUGS
.IX Header "BUGS"
.\"
Any bugs which are found should be reported on the project website:
.br
\fIhttps://support.endlessm.com/\fP
.\"
.SH AUTHOR
.IX Header "AUTHOR"
.\"
Endless Mobile, Inc.
.\"
.SH COPYRIGHT
.IX Header "COPYRIGHT"
.\"
Copyright © 2017 Endless Mobile, Inc.


debian/changelog
eos-updater (1.3.0~rc3) unstable; urgency=low

  * Updating version to 1.3.0~rc3

 -- Roddy Shuler <roddy@endlessm.com>  Tue, 12 Nov 2013 21:13:35 -0800

eos-updater (1.3.0~rc1) unstable; urgency=low

  * Updating version to 1.3.0~rc1

 -- Roddy Shuler <roddy@endlessm.com>  Fri, 08 Nov 2013 23:38:59 -0800

eos-updater (1.2.1~rc1) unstable; urgency=low

  * Updating version to 1.2.1~rc1

 -- Roddy Shuler <roddy@endlessm.com>  Sun, 27 Oct 2013 22:30:28 -0700

eos-updater (0-1) UNRELEASED; urgency=low

  * Initial release (Closes: #1144)

 -- John Hiesey <john@endlessm.com>  Fri, 04 Oct 2013 11:44:06 -0700


src/eos-autoupdater.c
/* -*- mode: C; c-file-style: "gnu"; indent-tabs-mode: nil; -*-
 *
 * Copyright © 2013, 2014, 2015, 2016, 2017 Endless Mobile, Inc.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

#include "eos-updater-types.h"
#include "eos-updater-generated.h"
#include <libeos-updater-util/config.h>

#include <gio/gio.h>
#include <glib.h>
#include <locale.h>
#include <stdlib.h>
#include <errno.h>

#include <nm-client.h>
#include <nm-device.h>

#include <systemd/sd-journal.h>

#define EOS_UPDATER_INVALID_ARGS_MSGID          "27b3a4600f7242acadf1855a2a1eaa6d"
#define EOS_UPDATER_CONFIGURATION_ERROR_MSGID   "5af9f4df37f949a1948971e00be0d620"
#define EOS_UPDATER_DAEMON_ERROR_MSGID          "f31fd043074a4a21b04784cf895c56ae"
#define EOS_UPDATER_STAMP_ERROR_MSGID           "da96f3494a5d432d8bcea1217433ecbf"
#define EOS_UPDATER_SUCCESS_MSGID               "ce0a80bb9f734dc09f8b56a7fb981ae4"
#define EOS_UPDATER_NOT_ONLINE_MSGID            "2797d0eaca084a9192e21838ab12cbd0"
#define EOS_UPDATER_MOBILE_CONNECTED_MSGID      "7c80d571cbc248d2a5cfd985c7cbd44c"
#define EOS_UPDATER_NOT_TIME_MSGID              "7c853d8fbc0b4a9b9f331b5b9aee4435"

/* The step of the update. These constants are used in the configuration
 * file to indicate which is the final automatic step before the user
 * needs to intervene.
 */
typedef enum _UpdateStep {
  UPDATE_STEP_NONE = 0,
  UPDATE_STEP_POLL = 1,
  UPDATE_STEP_FETCH = 2,
  UPDATE_STEP_APPLY = 3,
} UpdateStep;

/* These must be kept in sync with #UpdateStep. */
#define UPDATE_STEP_FIRST UPDATE_STEP_NONE
#define UPDATE_STEP_LAST UPDATE_STEP_APPLY

#define SEC_PER_DAY (3600ll * 24)

/* This file is touched whenever the updater starts */
static const char *UPDATE_STAMP_DIR = LOCALSTATEDIR "/lib/eos-updater";
static const char *UPDATE_STAMP_NAME = "eos-updater-stamp";

static const char *CONFIG_FILE_PATH = SYSCONFDIR "/" PACKAGE "/eos-autoupdater.conf";
static const char *OLD_CONFIG_FILE_PATH = SYSCONFDIR "/eos-updater.conf";
static const char *STATIC_CONFIG_FILE_PATH = PKGDATADIR "/eos-autoupdater.conf";
static const char *LOCAL_CONFIG_FILE_PATH = PREFIX "/local/share/" PACKAGE "/eos-autoupdater.conf";
static const char *AUTOMATIC_GROUP = "Automatic Updates";
static const char *LAST_STEP_KEY = "LastAutomaticStep";
static const char *INTERVAL_KEY = "IntervalDays";
static const char *ON_MOBILE_KEY = "UpdateOnMobile";

/* Ensures that the updater never tries to poll twice in one run */
static gboolean polled_already = FALSE;

/* Read from config file */
static UpdateStep last_automatic_step = UPDATE_STEP_NONE;

/* Set when main should return failure */
static gboolean should_exit_failure = FALSE;

/* Avoid erroneous additional state transitions */
static guint previous_state = EOS_UPDATER_STATE_NONE;

static GMainLoop *main_loop = NULL;
static gchar *volume_path = NULL;

static const gchar *
get_envvar_or (const gchar *envvar,
               const gchar *default_value)
{
  const gchar *value = g_getenv (envvar);

  if (value != NULL)
    return value;

  return default_value;
}

static const gchar *
get_stamp_dir (void)
{
  return get_envvar_or ("EOS_UPDATER_TEST_AUTOUPDATER_UPDATE_STAMP_DIR",
                        UPDATE_STAMP_DIR);
}

/* Note: This function does not report errors as a GError because there’s no
 * harm in the stamp file not being updated: it just means we’re going to check
 * again for updates sooner than otherwise. */
static void
update_stamp_file (void)
{
  const gchar *stamp_dir = get_stamp_dir ();
  g_autofree gchar *stamp_path = NULL;
  g_autoptr(GFile) stamp_file = NULL;
  g_autoptr(GError) error = NULL;

  if (g_mkdir_with_parents (stamp_dir, 0755) != 0) {
    int saved_errno = errno;
    const char *err_str = g_strerror (saved_errno);

    sd_journal_send ("MESSAGE_ID=%s", EOS_UPDATER_CONFIGURATION_ERROR_MSGID,
                     "PRIORITY=%d", LOG_CRIT,
                     "MESSAGE=Failed to create updater timestamp directory: %s", err_str,
                     NULL);
    return;
  }

  stamp_path = g_build_filename (stamp_dir, UPDATE_STAMP_NAME, NULL);
  stamp_file = g_file_new_for_path (stamp_path);
  g_file_replace_contents (stamp_file, "", 0, NULL, FALSE,
                           G_FILE_CREATE_NONE, NULL, NULL,
                           &error);
  if (error) {
    sd_journal_send ("MESSAGE_ID=%s", EOS_UPDATER_STAMP_ERROR_MSGID,
                     "PRIORITY=%d", LOG_CRIT,
                     "MESSAGE=Failed to write updater stamp file: %s", error->message,
                     NULL);
    return;
  }
}

/* Called on completion of the async dbus calls to check whether they
 * succeeded. Success doesn't mean that the operation succeeded, but it
 * does mean the call reached the daemon.
 */
static void
update_step_callback (GObject *source_object, GAsyncResult *res,
                      gpointer step_data)
{
  EosUpdater *proxy = (EosUpdater *) source_object;
  UpdateStep step = GPOINTER_TO_INT (step_data);
  GError *error = NULL;

  switch (step) {
    case UPDATE_STEP_POLL:
      eos_updater_call_poll_finish (proxy, res, &error);
      break;

    case UPDATE_STEP_FETCH:
      eos_updater_call_fetch_finish (proxy, res, &error);
      break;

    case UPDATE_STEP_APPLY:
      eos_updater_call_apply_finish (proxy, res, &error);
      break;

    case UPDATE_STEP_NONE:
    default:
      g_assert_not_reached ();
  }

  if (error) {
    sd_journal_send ("MESSAGE_ID=%s", EOS_UPDATER_DAEMON_ERROR_MSGID,
                     "PRIORITY=%d", LOG_ERR,
                     "MESSAGE=Error calling EOS updater: %s", error->message,
                     NULL);
    should_exit_failure = TRUE;
    g_main_loop_quit (main_loop);
    g_error_free (error);
  }
}

static gboolean
do_update_step (UpdateStep step, EosUpdater *proxy)
{
  gpointer step_data = GINT_TO_POINTER (step);

  /* Don't do more of the process than configured */
  if (step > last_automatic_step)
    return FALSE;

  switch (step) {
    case UPDATE_STEP_POLL:
      /* Don't poll more than once, or we will get stuck in a loop */
      if (polled_already)
        return FALSE;

      polled_already = TRUE;
      if (volume_path != NULL)
        eos_updater_call_poll_volume (proxy, volume_path, NULL, update_step_callback, step_data);
      else
        eos_updater_call_poll (proxy, NULL, update_step_callback, step_data);
      break;

    case UPDATE_STEP_FETCH:
      eos_updater_call_fetch (proxy, NULL, update_step_callback, step_data);
      break;

    case UPDATE_STEP_APPLY:
      eos_updater_call_apply (proxy, NULL, update_step_callback, step_data);
      break;

    case UPDATE_STEP_NONE:
    default:
      g_assert_not_reached ();
  }

  return TRUE;
}

static void
report_error_status (EosUpdater *proxy)
{
  const gchar *name, *error_message;

  name = eos_updater_get_error_name (proxy);
  error_message = eos_updater_get_error_message (proxy);

  sd_journal_send ("MESSAGE_ID=%s", EOS_UPDATER_DAEMON_ERROR_MSGID,
                   "PRIORITY=%d", LOG_ERR,
                   "MESSAGE=EOS updater error (%s): %s", name, error_message,
                   NULL);
}

/* The autoupdater is driven by state transitions in the updater daemon.
 * Whenever the state changes, we check if we need to do something as a
 * result of that state change. */
static void
on_state_changed (EosUpdater *proxy, EosUpdaterState state)
{
  gboolean continue_running = TRUE;

  if (state == previous_state)
    return;

  previous_state = state;

  g_message ("EOS updater state is: %u", state);

  switch (state) {
    case EOS_UPDATER_STATE_NONE: /* State should change soon */
      break;

    case EOS_UPDATER_STATE_READY: /* Must poll */
      continue_running = do_update_step (UPDATE_STEP_POLL, proxy);
      break;

    case EOS_UPDATER_STATE_ERROR: /* Log error and quit */
      report_error_status (proxy);
      should_exit_failure = TRUE;
      continue_running = FALSE;
      break;

    case EOS_UPDATER_STATE_POLLING: /* Wait for completion */
      break;

    case EOS_UPDATER_STATE_UPDATE_AVAILABLE: /* Possibly fetch */
      continue_running = do_update_step (UPDATE_STEP_FETCH, proxy);
      break;

    case EOS_UPDATER_STATE_FETCHING: /* Wait for completion */
      break;

    case EOS_UPDATER_STATE_UPDATE_READY: /* Possibly apply */
      continue_running = do_update_step (UPDATE_STEP_APPLY, proxy);
      break;

    case EOS_UPDATER_STATE_APPLYING_UPDATE: /* Wait for completion */
      break;

    case EOS_UPDATER_STATE_UPDATE_APPLIED: /* Done; exit */
      continue_running = FALSE;
      break;

    default:
      g_critical ("EOS updater entered invalid state: %u", state);
      continue_running = FALSE;
      should_exit_failure = TRUE;
      break;
  }

  if (!continue_running) {
    g_main_loop_quit (main_loop);
  }
}

static void
on_state_changed_notify (EosUpdater *proxy,
                         GParamSpec *pspec,
                         gpointer data)
{
  EosUpdaterState state = eos_updater_get_state (proxy);
  on_state_changed (proxy, state);
}

static const gchar *
get_config_file_path (void)
{
  return get_envvar_or ("EOS_UPDATER_TEST_AUTOUPDATER_CONFIG_FILE_PATH",
                        CONFIG_FILE_PATH);
}

static gboolean
read_config_file (const gchar *config_path,
                  guint *update_interval_days,
                  gboolean *update_on_mobile)
{
  g_autoptr(GKeyFile) config = NULL;
  g_autoptr(GError) error = NULL;
  gint _update_interval_days;
  gint _last_automatic_step;
  const gchar * const paths[] =
    {
      config_path,  /* typically CONFIG_FILE_PATH unless testing */
      OLD_CONFIG_FILE_PATH,
      LOCAL_CONFIG_FILE_PATH,
      STATIC_CONFIG_FILE_PATH,
      NULL
    };

  g_return_val_if_fail (update_interval_days != NULL, FALSE);
  g_return_val_if_fail (update_on_mobile != NULL, FALSE);

  /* Try loading the files in order */
  config = eos_updater_load_config_file (paths, &error);

  if (error != NULL) {
    sd_journal_send ("MESSAGE_ID=%s", EOS_UPDATER_CONFIGURATION_ERROR_MSGID,
                     "PRIORITY=%d", LOG_ERR,
                     "MESSAGE=Unable to open the configuration file: %s", error->message,
                     NULL);
    return FALSE;
  }

  /* Successfully loaded a file. Parse it. */
  _last_automatic_step = g_key_file_get_integer (config, AUTOMATIC_GROUP,
                                                 LAST_STEP_KEY, &error);
  if (error) {
    sd_journal_send ("MESSAGE_ID=%s", EOS_UPDATER_CONFIGURATION_ERROR_MSGID,
                     "PRIORITY=%d", LOG_ERR,
                     "MESSAGE=Unable to read key '%s' in config file", LAST_STEP_KEY,
                     NULL);
    return FALSE;
  }

  if (_last_automatic_step < UPDATE_STEP_FIRST ||
      _last_automatic_step > UPDATE_STEP_LAST) {
    sd_journal_send ("MESSAGE_ID=%s", EOS_UPDATER_CONFIGURATION_ERROR_MSGID,
                     "PRIORITY=%d", LOG_ERR,
                     "MESSAGE=Specified last automatic step is not a valid step",
                     NULL);
    return FALSE;
  }

  last_automatic_step = (UpdateStep) _last_automatic_step;

  _update_interval_days = g_key_file_get_integer (config, AUTOMATIC_GROUP,
                                                  INTERVAL_KEY, &error);

  if (error) {
    sd_journal_send ("MESSAGE_ID=%s", EOS_UPDATER_CONFIGURATION_ERROR_MSGID,
                     "PRIORITY=%d", LOG_ERR,
                     "MESSAGE=Unable to read key '%s' in config file", INTERVAL_KEY,
                     NULL);
    return FALSE;
  }

  if (_update_interval_days < 0) {
    sd_journal_send ("MESSAGE_ID=%s", EOS_UPDATER_CONFIGURATION_ERROR_MSGID,
                     "PRIORITY=%d", LOG_ERR,
                     "MESSAGE=Specified update interval is less than zero",
                     NULL);
    return FALSE;
  }

  /* This should always be true, as the RHS is out of range for a guint (it’s
   * around 10^14 days, which should be a long enough update period for anyone).
   * We use G_MAXUINT64 rather than G_MAXUINT because the time calculation in
   * is_time_to_update() uses guint64 variables. */
  g_assert ((guint64) _update_interval_days <= G_MAXUINT64 / SEC_PER_DAY);

  *update_interval_days = (guint) _update_interval_days;

  *update_on_mobile = g_key_file_get_boolean (config, AUTOMATIC_GROUP,
                                              ON_MOBILE_KEY, &error);

  if (error) {
    sd_journal_send ("MESSAGE_ID=%s", EOS_UPDATER_CONFIGURATION_ERROR_MSGID,
                     "PRIORITY=%d", LOG_ERR,
                     "MESSAGE=Unable to read key '%s' in config file", ON_MOBILE_KEY,
                     NULL);
    return FALSE;
  }

  return TRUE;
}

/* We want to poll once when the updater starts.  To make sure that we
 * can quit ourselves gracefully, we wait until the main loop starts.
 */
static gboolean
initial_poll_idle_func (gpointer pointer)
{
  EosUpdater *proxy = (EosUpdater *) pointer;
  EosUpdaterState initial_state = eos_updater_get_state (proxy);

  /* Attempt to clear the error by pretending to be ready, which will
   * trigger a poll
   */
  if (initial_state == EOS_UPDATER_STATE_ERROR)
    initial_state = EOS_UPDATER_STATE_READY;

  on_state_changed (proxy, initial_state);

  /* Disable this function after the first run */
  return G_SOURCE_REMOVE;
}

static gboolean
is_time_to_update (guint update_interval_days)
{
  const gchar *stamp_dir = get_stamp_dir ();
  g_autofree gchar *stamp_path = NULL;
  g_autoptr (GFile) stamp_file = NULL;
  g_autoptr (GFileInfo) stamp_file_info = NULL;
  guint64 last_update_time_secs;
  gint64 current_time_usec;
  g_autoptr (GError) error = NULL;
  gboolean is_time_to_update = FALSE;

  stamp_path = g_build_filename (stamp_dir, UPDATE_STAMP_NAME, NULL);
  stamp_file = g_file_new_for_path (stamp_path);
  stamp_file_info = g_file_query_info (stamp_file,
                                       G_FILE_ATTRIBUTE_TIME_MODIFIED,
                                       G_FILE_QUERY_INFO_NONE, NULL,
                                       &error);

  if (error) {
    if (!g_error_matches (error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND)) {
      /* Failed for some reason other than the file not being present */
      sd_journal_send ("MESSAGE_ID=%s", EOS_UPDATER_STAMP_ERROR_MSGID,
                       "PRIORITY=%d", LOG_CRIT,
                       "MESSAGE=Failed to read attributes of updater timestamp file",
                       NULL);
    }

    is_time_to_update = TRUE;
  } else {
    guint64 next_update_time_secs, update_interval_secs;

    /* Determine whether sufficient time has elapsed */
    current_time_usec = g_get_real_time ();
    last_update_time_secs =
      g_file_info_get_attribute_uint64 (stamp_file_info,
                                        G_FILE_ATTRIBUTE_TIME_MODIFIED);

    /* Guaranteed not to overflow, as we check update_interval_days when
     * loading it. */
    update_interval_secs = update_interval_days * SEC_PER_DAY;

    /* next_update_time_secs = last_update_time_secs + update_interval_secs */
    if (!g_uint64_checked_add (&next_update_time_secs, last_update_time_secs,
                               update_interval_secs))
      next_update_time_secs = G_MAXUINT64;

    is_time_to_update = (next_update_time_secs < (guint64) current_time_usec / G_USEC_PER_SEC);
  }

  return is_time_to_update;
}

static gboolean
should_listen_on_session_bus (void)
{
  const gchar *value = NULL;

  value = g_getenv ("EOS_UPDATER_TEST_AUTOUPDATER_USE_SESSION_BUS");

  return value != NULL;
}

static gboolean
is_online (void)
{
  NMClient *client;
  gboolean online;

  /* Don’t connect to NetworkManager when we are supposed to use the session
   * bus, as NM is on the system bus, and we don’t want to mock it up. */
  if (should_listen_on_session_bus ())
    {
      g_message ("Not using NetworkManager: assuming network is online.");
      return TRUE;
    }

  client = nm_client_new ();
  if (!client)
    return FALSE;

  /* Assume that the ostree server is remote and only consider to be
   * online for ostree updates if we have global connectivity.
   * For Avahi updates, local or site connectivity is adequate.
   */
  switch (nm_client_get_state (client)) {
  case NM_STATE_CONNECTED_LOCAL:
  case NM_STATE_CONNECTED_SITE:
  case NM_STATE_CONNECTED_GLOBAL:
    online = TRUE;
    break;
  case NM_STATE_UNKNOWN:
  case NM_STATE_ASLEEP:
  case NM_STATE_DISCONNECTED:
  case NM_STATE_DISCONNECTING:
  case NM_STATE_CONNECTING:
  default:
    online = FALSE;
    break;
  }
  g_object_unref (client);

  if (!online)
    sd_journal_send ("MESSAGE_ID=%s", EOS_UPDATER_NOT_ONLINE_MSGID,
                     "PRIORITY=%d", LOG_INFO,
                     "MESSAGE=Not currently online. Not updating",
                     NULL);
  return online;
}

static gboolean
is_connected_through_mobile (void)
{
  NMActiveConnection *connection;
  NMClient *client;
  NMDevice *device;
  const GPtrArray *devices;
  gboolean is_mobile = FALSE;
  guint i;

  /* Don’t connect to NetworkManager when we are supposed to use the session
   * bus, as NM is on the system bus, and we don’t want to mock it up. */
  if (should_listen_on_session_bus ())
    {
      g_message ("Not using NetworkManager: assuming network is not mobile.");
      return FALSE;
    }

  client = nm_client_new ();
  if (!client) {
    return FALSE;
  }

  connection = nm_client_get_primary_connection (client);
  if (!connection) {
    g_object_unref (client);
    return FALSE;
  }

  devices = nm_active_connection_get_devices (connection);
  for (i = 0; !is_mobile && i < devices->len; i++) {
    device = (NMDevice *) g_ptr_array_index (devices, i);
    switch (nm_device_get_device_type (device)) {
    case NM_DEVICE_TYPE_MODEM:
    case NM_DEVICE_TYPE_BT:
    case NM_DEVICE_TYPE_WIMAX:
      is_mobile = TRUE;
      break;
    case NM_DEVICE_TYPE_UNKNOWN:
    case NM_DEVICE_TYPE_ETHERNET:
    case NM_DEVICE_TYPE_WIFI:
    case NM_DEVICE_TYPE_UNUSED1:
    case NM_DEVICE_TYPE_UNUSED2:
    case NM_DEVICE_TYPE_OLPC_MESH:
    case NM_DEVICE_TYPE_INFINIBAND:
    case NM_DEVICE_TYPE_BOND:
    case NM_DEVICE_TYPE_VLAN:
    case NM_DEVICE_TYPE_ADSL:
    case NM_DEVICE_TYPE_BRIDGE:
    case NM_DEVICE_TYPE_GENERIC:
    case NM_DEVICE_TYPE_TEAM:
    case NM_DEVICE_TYPE_TUN:
    case NM_DEVICE_TYPE_IP_TUNNEL:
    case NM_DEVICE_TYPE_MACVLAN:
    case NM_DEVICE_TYPE_VXLAN:
    case NM_DEVICE_TYPE_VETH:
    default:
      break;
    }
  }

  g_object_unref (client);

  return is_mobile;
}

static gint
get_dbus_timeout (void)
{
  const gchar *value = NULL;
  gint64 timeout;
  gchar *str_end = NULL;

  value = get_envvar_or ("EOS_UPDATER_TEST_AUTOUPDATER_DBUS_TIMEOUT",
                         NULL);

  if (value == NULL || value[0] == '\0')
    return -1;

  errno = 0;
  timeout = g_ascii_strtoll (value, &str_end, 10);
  if (errno != 0 ||
      str_end == NULL ||
      *str_end != '\0' ||
      timeout > G_MAXINT ||
      timeout < 0)
    return -1;

  return timeout;
}

/* main() exit codes. */
enum
{
  EXIT_OK = EXIT_SUCCESS,
  EXIT_FAILED = 1,
  EXIT_INVALID_ARGUMENTS = 2,
  EXIT_BAD_CONFIGURATION = 3,
};

int
main (int argc, char **argv)
{
  g_autoptr(EosUpdater) proxy = NULL;
  g_autoptr(GError) error = NULL;
  guint update_interval_days;
  gboolean update_on_mobile;
  gboolean force_update = FALSE;
  g_autoptr(GOptionContext) context = NULL;

  GOptionEntry entries[] = {
    { "force-update", 0, 0, G_OPTION_ARG_NONE, &force_update, "Force an update", NULL },
    { "from-volume", 0, 0, G_OPTION_ARG_STRING, &volume_path, "Poll for updates from the volume", "PATH" },
    { NULL }
  };
  GBusType bus_type = G_BUS_TYPE_SYSTEM;
  gint dbus_timeout;

  setlocale (LC_ALL, "");

  context = g_option_context_new ("— Endless OS Automatic Updater");
  g_option_context_add_main_entries (context, entries, NULL);
  g_option_context_set_summary (context,
                                "Automatically poll for, fetch and apply "
                                "updates in the background. This drives the "
                                "state changes in the eos-updater service.");

  if (!g_option_context_parse (context, &argc, &argv, &error))
    {
      sd_journal_send ("MESSAGE_ID=%s", EOS_UPDATER_INVALID_ARGS_MSGID,
                       "PRIORITY=%d", LOG_ERR,
                       "MESSAGE=Error parsing command line arguments: %s", error->message,
                       NULL);
      return EXIT_INVALID_ARGUMENTS;
    }

  if (!read_config_file (get_config_file_path (),
                         &update_interval_days, &update_on_mobile))
    return EXIT_BAD_CONFIGURATION;

  if (volume_path == NULL && !is_online ())
    return EXIT_OK;

  /* Always force an update if running with --from-volume; it doesn’t make
   * sense not to. */
  if (volume_path != NULL)
    force_update = TRUE;

  if (!force_update) {
    if (volume_path == NULL &&
        !update_on_mobile &&
        is_connected_through_mobile ()) {
      sd_journal_send ("MESSAGE_ID=%s", EOS_UPDATER_MOBILE_CONNECTED_MSGID,
                       "PRIORITY=%d", LOG_INFO,
                       "MESSAGE=Connected to mobile network. Not updating",
                       NULL);
      return EXIT_OK;
    }

    if (!is_time_to_update (update_interval_days)) {
      sd_journal_send ("MESSAGE_ID=%s", EOS_UPDATER_NOT_TIME_MSGID,
                       "PRIORITY=%d", LOG_INFO,
                       "MESSAGE=Less than %s since last update. Exiting", INTERVAL_KEY,
                       NULL);
      return EXIT_OK;
    }
  }

  main_loop = g_main_loop_new (NULL, FALSE);

  if (should_listen_on_session_bus ())
    bus_type = G_BUS_TYPE_SESSION;
  proxy = eos_updater_proxy_new_for_bus_sync (bus_type,
                     G_DBUS_PROXY_FLAGS_NONE,
                     "com.endlessm.Updater",
                     "/com/endlessm/Updater",
                     NULL,
                     &error);

  if (error) {
    sd_journal_send ("MESSAGE_ID=%s", EOS_UPDATER_DAEMON_ERROR_MSGID,
                     "PRIORITY=%d", LOG_ERR,
                     "MESSAGE=Error getting EOS updater object: %s", error->message,
                     NULL);
    should_exit_failure = TRUE;
    goto out;
  }

  dbus_timeout = get_dbus_timeout ();
  g_dbus_proxy_set_default_timeout (G_DBUS_PROXY (proxy), dbus_timeout);

  g_signal_connect (proxy, "notify::state",
                    G_CALLBACK (on_state_changed_notify), NULL);

  g_idle_add (initial_poll_idle_func, proxy);
  g_main_loop_run (main_loop);

out:
  g_main_loop_unref (main_loop);
  g_free (volume_path);

  if (should_exit_failure) /* All paths setting this print an error message */
    return EXIT_FAILED;

  /* Update the stamp file since all configured steps have succeeded. */
  update_stamp_file ();
  sd_journal_send ("MESSAGE_ID=%s", EOS_UPDATER_SUCCESS_MSGID,
                   "PRIORITY=%d", LOG_INFO,
                   "MESSAGE=Updater finished successfully",
                   NULL);

  return EXIT_OK;
}
